python/torch_mlir/extras/fx_importer.py

```diff
@@ -1485,6 +1585,74 @@ def _import_torch_op_overload(
         for i, value in enumerate(operation.results):
             self.bind_node_value(node, value, i)
 
+    def _import_shape_symbols_with_guards(
+        self, loc: Location, symbolic_guards: Dict[str, RangeConstraint]
+    ):
+        for symbol, constraints in symbolic_guards.items():
+            # Create torch.sym_int ops
+            operation = Operation.create(
+                name="torch.symbolic_int",
+                attributes={
+                    "symbol_name": StringAttr.get(symbol),
+                    "min_val": self._cc.integer_attr(constraints.min_val, 64),
+                    "max_val": self._cc.integer_attr(constraints.max_val, 64),
+                },
+                results=[self._cc.torch_int_type],
+                loc=loc,
+            )
+            self.bind_symbol_value(symbol, operation.result)
+
+    def _create_bind_symbolic_shape_ops(self, loc: Location, node: torch_fx.Node):
+        node_val = node.meta.get("val")
+        if (node_val is not None) and isinstance(node_val, TorchFakeTensor):
+            # Only create bind ops if the shapes contain symbolic sizes
+            if node_val._has_symbolic_sizes_strides:
+                # Read node metadata to obtain shape symbols and expressions
+                symbols_set = set()
+                shape_exprs = []
+                for s in node_val.size():
+                    if isinstance(s, torch.SymInt):
+                        symbols_set.update(s.node.expr.free_symbols)
+                        shape_exprs.append(s.node.expr)
+                    else:
+                        assert isinstance(s, int)
+                        shape_exprs.append(s)
+
+                # Map from actual shape symbols to local symbols in the affine map
+                symbols_set = sorted(symbols_set, key=lambda x: x.name)
+
+                symbol_table = {
+                    str(symbol): AffineSymbolExpr.get(i)
+                    for i, symbol in enumerate(symbols_set)
+                }
+
+                # Convert symbolic shape expressions into affine expressions
+                affine_exprs = [
+                    (
+                        AffineConstantExpr.get(expr)
+                        if isinstance(expr, int)
+                        # Use `symbol_table.copy()` here to avoid `eval` from
+                        # adding the `__builtins__` entry to the map.
+                        else eval(str(expr), symbol_table.copy())
```

Comment:
Using eval like this isn't great. Just going to leave this comment for a second while considering.

Reply:
Good callout. I'm revamping this to actually recursively traverse the sympy AST and translate it to AffineExpr depending on the type of sympy expr (Symbol vs Add vs Pow etc). Encountered a case of `s0**2` (sympy.Pow) which AffineExpr don't support natively (i.e. `s0**2` isn't but `s0 * s0` is supported). I think most such cases should be handled with the `mul`, `floordiv` , `ceildiv` and `mod` from semi-affine maps : https://mlir.llvm.org/docs/Dialects/Affine/#semi-affine-maps.